iceoryx_posh/test/moduletests/test_popo_listener_waitset_attachments.cpp

```diff
@@ -0,0 +1,210 @@
+using namespace ::iox::capro;
+using namespace ::iox::mepoo;
+
+constexpr const char RUNTIME_NAME[] = "torben_dallas";
+constexpr const char SERVICE[] = "respect_to_the_man_in_the_icecream_van";
+constexpr const char INSTANCE[] = "Lakierski materialski";
+constexpr const char EVENT[] = "boom boom boomerang";
+
+/// Those tests verify that the destructor of the attachables is correct. When we
+/// use inheritance it is possible that the trigger is a member of the base class
+/// and destroyed in the base class constructor as well.
+/// But the listener and waitset require the child class for cleanup. If the child
+/// class destructor does not call reset on all the triggers the base class will
+/// and the listener and waitset will call a method on the original class which is
+/// already deleted. This causes undefined behavior which the sanitizer will catch.
+///
+/// The following tests should be run for all classes which can be attached to a
+/// listener or a waitset to ensure that trigger.reset() is called in the child destructor
+/// when we work with inheritance.
+///
+/// When no inheritance is used we shouldn't encounter any problems.
+///
+/// It suffices to call all tests on the listener since the listener and the waitset
+/// are using the same trigger concept
+class ListenerWaitsetAttachments_test : public Test
+{
+  public:
+    void SetUp()
+    {
+        EXPECT_CALL(*this->runtimeMock, getMiddlewareConditionVariable())
+            .WillOnce(Return(&this->conditionVariableData));
+        listener.emplace();
+    }
+
+    template <typename T>
+    static void genericTriggerCallback(T* const)
+    {
+    }
+
+    std::unique_ptr<PoshRuntimeMock> runtimeMock = PoshRuntimeMock::create(RUNTIME_NAME);
+    ConditionVariableData conditionVariableData{RUNTIME_NAME};
+    optional<Listener> listener;
+
+    MemoryManager memoryManager;
+};
+
+
+TEST_F(ListenerWaitsetAttachments_test, SubscriberAttachesToListener)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "f1d163e0-4479-47b4-b4e8-01b0ee6a71b0");
+    SubscriberPortData subscriberData({SERVICE, INSTANCE, EVENT},
+                                      RUNTIME_NAME,
+                                      VariantQueueTypes::SoFi_MultiProducerSingleConsumer,
+                                      SubscriberOptions());
+    EXPECT_CALL(*this->runtimeMock, getMiddlewareSubscriber(_, _, _)).WillOnce(Return(&subscriberData));
+
+    optional<Subscriber<int>> subscriber;
+    subscriber.emplace(ServiceDescription(SERVICE, INSTANCE, EVENT));
+
+    ASSERT_FALSE(listener
+                     ->attachEvent(*subscriber,
+                                   SubscriberEvent::DATA_RECEIVED,
+                                   createNotificationCallback(
+                                       ListenerWaitsetAttachments_test::genericTriggerCallback<Subscriber<int>>))
+                     .has_error());
+
+    EXPECT_THAT(listener->size(), Eq(1));
+    subscriber.reset();
```

Comment:
reset causes it to be detached, is this what we want to check here (then the test name should be adapted).

Same for all the other tests and maybe some explanatory comment. The comment at the beginning of the tests is not fully clear to me.

Reply:
Yes, the idea is to reset the subscriber before the Listener is destroyed to check for undefined behavior in the CI

Reply:
OK, we still cannot be sure it actually calls the trigger reset (as the test name states) I think (could it be detached due to another reason?)I guess it fine as an UB check.